docs/madr/decisions/040-meshservice-hostname-vips.md

```diff
@@ -0,0 +1,127 @@
+# Hostnames and VIP management in MeshService
+
+* Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/6331
+
+## Context and Problem Statement
+
+With introduction of MeshService, we need to have a mechanism to assign VIPs and hostnames to MeshServices.
+Currently, VIPs are managed in a ConfigMap per Mesh. Hostnames are managed using VirtualOutbound.
+
+## Considered Options
+
+* VIPs management
+* Hostname Generator
+* Static Hostname template in CP config
+
+## Decision Outcome
+
+Chosen option: "Hostname Generator" and "VIPs management" since it's the most flexible option.
+
+## Pros and Cons of the Options
+
+### VIPs management
+
+We have to assign a VIP for each MeshService to be able to address it using transparent proxy.
+
+MeshService object contains VIPs in status.
+```yaml
+status: # managed by CP. Not shared cross zone, but synced to global
+  vip:
+    ip: <kube_cluster_ip> # or kuma VIP or Pod IP for Headless Service
+    type: Kubernetes # | Kuma
+```
+
+If MeshService was created based on Kubernetes Service, we can just take ClusterIP, but otherwise we need to assign each MeshService a Kuma VIP.
+Kuma VIP is a non-routable IP used only by Envoy to indicate destination (like a host header, but on L4). We currently do this with 240.0.0.0/4 CIDR.
+
+Kuma VIPs are assigned asynchronously with MeshService creation. Assigning Kuma VIPs as sync operation on MeshService creation would create a race condition on VIP allocation.
+
+VIP assigning process (single-threaded):
```

Comment:
is it a controller on k8s?

Reply:
I'd say it's Kuma component with a ticker. We need to aggregate all MeshServices to compute this.